frame/society/src/lib.rs

```diff
@@ -0,0 +1,894 @@
+
+					let strikes = <Strikes<T, I>>::mutate(m, |s| {
+						*s += 1;
+						*s
+					});
+					if strikes >= T::MaxStrikes::get() {
+						Self::suspend_member(m);
+					}
+					amount
+				};
+
+				// Collect the voters who had a matching vote.
+				rewardees.extend(votes.into_iter()
+					.filter_map(|(v, m)|
+						if v == matching_vote { Some(m) } else {
+							total_slash += bad_vote(m);
+							None
+						}
+					).cloned()
+				);
+
+				if is_accepted {
+					total_approvals += approval_count;
+					total_payouts += value;
+					members.push(candidate.clone());
+
+					Self::pay_accepted_candidate(&candidate, value, kind, maturity);
+
+					// We track here the total_approvals so that every candidate has a unique range
+					// of numbers from 0 to `total_approvals` with length `approval_count` so each
+					// candidate is proportionally represented when selecting a "primary" below.
+					Some((candidate, total_approvals))
+				} else {
+					// Suspend Candidate
+					<SuspendedCandidates<T, I>>::insert(&candidate, (value, kind));
+					Self::deposit_event(RawEvent::CandidateSuspended(candidate));
+					None
+				}
+			}).collect::<Vec<_>>();
+
+			// Reward one of the voters who voted the right way.
+			if !total_slash.is_zero() {
+				if let Some(winner) = pick_item(&mut rng, &rewardees) {
+					// If we can't reward them, not much that can be done.
+					Self::bump_payout(winner, maturity, total_slash);
+				} else {
+					// Move the slashed amount back from payouts account to local treasury.
+					let _ = T::Currency::transfer(&Self::payouts(), &Self::account_id(), total_slash, AllowDeath);
+				}
+			}
+
+			// Fund the total payouts from the local treasury.
+			if !total_payouts.is_zero() {
+				// remove payout from pot and shift needed funds to the payout account.
+				pot = pot.saturating_sub(total_payouts);
+
+				// this should never fail since we ensure we can afford the payouts in a previous
+				// block, but there's not much we can do to recover if it fails anyway.
+				let _ = T::Currency::transfer(&Self::account_id(), &Self::payouts(), total_payouts, AllowDeath);
+			}
+
+			// if at least one candidate was accepted...
+			if !accepted.is_empty() {
+				// select one as primary, randomly chosen from the accepted, weighted by approvals. 
+				// Choose a random number between 0 and `total_approvals`
+				let primary_point = pick_usize(&mut rng, total_approvals - 1);
+				// Find the user who falls on that point
+				let primary = accepted.iter().find(|e| e.1 > primary_point)
+					.expect("e.1 of final item == total_approvals; \
+						worst case find will always return that item; qed")
+					.0.clone();
+				
+				let accounts = accepted.into_iter().map(|x| x.0).collect::<Vec<_>>();
+
+				// Then write everything back out, signal the changed membership and leave an event.
+				members.sort();
+				<Members<T, I>>::put(members.clone());
```

Comment:
This might be really stupid. I don't understand the relative costs between cloning a large vec like `members` versus reading it twice.

It seems I would need to do one of these two choices. Please let me know which is right.

Reply:
can't the following works 
```suggestion
				<Members<T, I>>::put(&members);
```
With EncodeLike mecanism you should be able to avoid most clone to write to storage now

Reply:
I get the following error

```
error[E0277]: the trait bound `&&mut std::vec::Vec<<T as frame_system::Trait>::AccountId>: _::_parity_scale_codec::EncodeLike<std::vec::Vec<<T as frame_system::Trait>::AccountId>>` is not satisfied
   --> frame/society/src/lib.rs:702:26
    |
702 |                 <Members<T, I>>::put(&members);
    |                                      -^^^^^^^
    |                                      |
    |                                      the trait `_::_parity_scale_codec::EncodeLike<std::vec::Vec<<T as frame_system::Trait>::AccountId>>` is not implemented for `&&mut std::vec::Vec<<T as frame_system::Trait>::AccountId>`
    |                                      help: consider removing 1 leading `&`-references
    |
    = help: the following implementations were found:
              <std::vec::Vec<T> as _::_parity_scale_codec::EncodeLike<&[U]>>
              <std::vec::Vec<T> as _::_parity_scale_codec::EncodeLike<std::vec::Vec<U>>>
    = note: required by `sp_api_hidden_includes_decl_storage::hidden_include::StorageValue::put`

Reply:
hmm yes ok, indeed you can do
```suggestion
				<Members<T, I>>::put(codec::Ref::from(&members));
```

or 
```suggestion
				<Members<T, I>>::put(&members[..]);
```